File: go/problem/has_cycle.go
package problem

func hasCycle(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return false
	}
	fast, slow := head, head
	for fast != nil {
		if fast.Next != nil {
			fast = fast.Next.Next
		} else {
			return false
		}
		slow = slow.Next

		if fast == slow {
			return true
		}
	}

	return false
}

File: rust/src/problems/has_cycle.rs
pub struct Solution;
use super::structs::ListNode;

impl Solution {
    pub fn has_cycle(head: Option<Box<ListNode>>) -> bool {
        let (mut slow, mut fast) = (head.as_ref(), head.as_ref());
        while let (Some(slow_node), Some(fast_node)) = (slow, fast) {
            slow = slow_node.next.as_ref();
            fast = fast_node.next.as_ref().and_then(|n| n.next.as_ref());

            if slow == fast && slow.is_some() {
                return true;
            }
        }
        return false;
    }
}
